File: java/ZeroMatrix.java
/*
 * Title: Zero Matrix
 * Author: Sam Gavis-Hughson
 * Date: 9/5/2016
 * 
 * Given a boolean matrix, update it so that if any cell is true, all the rows 
 * in that cell are true.
 * 
 * eg. 
 * [true,  false, false]      [true,  true,  true ]
 * [false, false, false]  ->  [true,  false, false]
 * [false, false, false]      [true,  false, false]
 * 
 * Execution: javac ZeroMatrix.java && java ZeroMatrix
 * 
 * For more details, check out http://www.byte-by-byte.com/zeromatrix/
 */
import java.util.Arrays;

public class ZeroMatrix {
    public static void zeroMatrix(boolean[][] matrix) {
        
        // Verify the input array is nonzero
        if (matrix.length == 0 || matrix[0].length == 0) return;
        
        // Determine whether the first row or first column is true
        boolean rowZero = false, colZero = false;
        for (boolean i : matrix[0]) {
            rowZero |= i;
        }
        for (boolean[] i : matrix) {
            colZero |= i[0];
        }
            
        // For each cell not in the first row/column, if it is true, set the 
        // cell in the first row/same column and first column/same row to be 
        // true
        for (int i = 1; i < matrix.length; i++) {
            for (int j = 1; j < matrix[0].length; j++) {
                if (matrix[i][j]) {
                    matrix[i][0] = true;
                    matrix[0][j] = true;
                }
            }
        }
        
        // Go through the first column and set each row to true where cell in
        // the first column is true
        for (int i = 1; i < matrix.length; i++) {
            if (matrix[i][0]) {
                for (int j = 1; j < matrix[i].length; j++) {
                    matrix[i][j] = true;
                }
            }
        }
        
        // Repeat for the rows
        for (int j = 1; j < matrix[0].length; j++) {
            if (matrix[0][j]) {
                for (int i = 1; i < matrix.length; i++) {
                    matrix[i][j] = true;
                }
            }
        }
        
        // Set first row/column to true if necessary
        if (rowZero) {
            for (int i = 0; i < matrix[0].length; i++) {
                matrix[0][i] = true;
            }
        }
        
        if (colZero) {
            for (int i = 0; i < matrix.length; i++) {
                matrix[i][0] = true;
            }
        }
    }
    
    // Sample test cases
    public static void main(String[] args) {
        boolean[][] a = new boolean[][]{
            {true, false, false}, 
            {false, false, false}, 
            {false, false, false}
        };
        zeroMatrix(a);
        assert compare2dArrays(a, new boolean[][]{
            {true, true, true},
            {true, false, false},
            {true, false, false}
        }) : "First row and first column";
        
        a = new boolean[][]{
            {false, false, false}, 
            {false, false, false}, 
            {false, false, false}
        };
        zeroMatrix(a);
        assert compare2dArrays(a, new boolean[][]{
            {false, false, false}, 
            {false, false, false}, 
            {false, false, false}
        }) : "All false";
        
        a = new boolean[][]{
            {true, true, true}, 
            {true, true, true}, 
            {true, true, true}, 
        };
        zeroMatrix(a);
        assert compare2dArrays(a, new boolean[][]{
            {true, true, true}, 
            {true, true, true}, 
            {true, true, true}, 
        }) : "All true";
        
        a = new boolean[][]{
            {false, true, false}, 
        };
        zeroMatrix(a);
        assert compare2dArrays(a, new boolean[][]{
            {true, true, true}
        }) : "1 x 3 array";
        
        a = new boolean[][]{
            {false}, 
            {true}, 
            {false}
        };
        zeroMatrix(a);
        assert compare2dArrays(a, new boolean[][]{
            {true}, 
            {true}, 
            {true}
        }) : "3 x 1 array";
        
        a = new boolean[][]{
            {true, false, false}, 
            {false, false, false}, 
            {false, false, true}
        };
        zeroMatrix(a);
        assert compare2dArrays(a, new boolean[][]{
            {true, true, true}, 
            {true, false, true}, 
            {true, true, true}
        }) : "Multiple true";
        
        System.out.println("Passed all test cases");
    }
    
    private static boolean compare2dArrays(boolean[][] a, boolean[][] b) {
        if (a.length != b.length) return false;
        if (a[0].length != b[0].length) return false;
        
        for (int i = 0; i < a.length; i++) {
            for (int j = 0; j < a[0].length; j++) {
                if (a[i][j] != b[i][j]) return false;
            }
        }
        
        return true;
    }
}

File: python/ZeroMatrix.py
"""
Title: Zero Matrix

Given a boolean matrix, update it so that if any cell is true, all the rows 
in that cell are true.

eg. 
[true,  false, false]      [true,  true,  true ]
[false, false, false]  ->  [true,  false, false]
[false, false, false]      [true,  false, false]
 
Execution: python ZeroMatrix.py
 
For more details, check out http://www.byte-by-byte.com/zeromatrix/
"""
import unittest


def zero_matrix(mat):
    row_flag = False
    col_flag = False

    for i in range(len(mat)):
        for j in range(len(mat)):
            if i == 0 and mat[i][j]:
                row_flag = True
            if j == 0 and mat[i][j]:
                col_flag = True
            if mat[i][j]:
                mat[0][j] = True
                mat[i][0] = True
    for i in range(1, len(mat)):
        for j in range(1, len(mat)):
            if mat[0][j] or mat[i][0]:
                mat[i][j] = True
    if row_flag:
        for i in range(len(mat)):
            mat[0][i] = True
    if col_flag:
        for i in range(len(mat)):
            mat[i][0] = True
    return mat


class TestBigIntMod(unittest.TestCase):
    def test_first_row_first_col(self):
        A = [[True, False, False],
             [False, False, False], 
             [False, False, False]]

        B = [[True, True, True],
             [True, False, False],
             [True, False, False]]

        self.assertEqual(zero_matrix(A), B)
        print("First row and first column")

    def test_all_false(self):
        A = [[False, False, False],
             [False, False, False], 
             [False, False, False]]

        B = [[False, False, False],
             [False, False, False],
             [False, False, False]]

        self.assertEqual(zero_matrix(A), B)
        print("All false.")

    def test_all_true(self):
        A = [[True, True, True],
             [True, True, True],
             [True, True, True]]

        B = [[True, True, True],
             [True, True, True],
             [True, True, True]]

        self.assertEqual(zero_matrix(A), B)
        print("All true.")

    def test_multiple_true(self):
        A = [[True, False, False],
             [False, False, False],
             [False, False, True]]

        B = [[True, True, True],
             [True, False, True],
             [True, True, True]]

        self.assertEqual(zero_matrix(A), B)
        print("Multiple true.")


if __name__ == '__main__':
    unittest.main()

